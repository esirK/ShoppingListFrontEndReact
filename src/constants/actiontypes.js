export const SUBMITTING_STARTED = 'SUBMITTING_STARTED';
export const GETTING_SHOPPINGLISTS_STARTED = 'GETTING_SHOPPINGLISTS_STARTED';
export const SUBMISSION_SUCCESSFUL = 'SUBMISSION_SUCCESSFUL';
export const SUBMISSION_ERROR = 'SUBMISSION_ERROR';
export const CLEAR_ERRORS = 'CLEAR_ERRORS';
export const SWITCH_USER_AUTH_STATUS = 'SWITCH_USER_AUTH_STATUS';
export const ERROR_ENCOUNTERED = 'ERROR_ENCOUNTERED';
export const SHOPPINGLISTS_LOADED_SUCCESSFULLY = 'SHOPPINGLISTS_LOADED_SUCCESSFULLY';
export const CREATING_SHOPPINGLISTS_STARTED = 'CREATING_SHOPPINGLISTS_STARTED';
export const SHOPPINGLISTS_CREATED_SUCCESSFULY = 'SHOPPINGLISTS_CREATED_SUCCESSFULY';
export const SHOPPINGLISTS_DELETED_SUCCESSFULY = 'SHOPPINGLISTS_DELETED_SUCCESSFULY';
export const SHOPPINGLISTS_UPDATED_SUCCESSFULY = 'SHOPPINGLISTS_UPDATED_SUCCESSFULY';