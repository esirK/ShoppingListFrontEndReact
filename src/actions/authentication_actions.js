import axios from 'axios';
import jwt from 'jsonwebtoken';

import * as types from '../constants/actiontypes';

import {URL, startSubmitting} from './index';

export function submissionSuccessful(message){
	return {
		type: types.SUBMISSION_SUCCESSFUL,
		message
	};
}
export function submissionFailed(error) {
	return{
		type: types.SUBMISSION_ERROR,
		error
	};
}
export function setAuthStatusOfUser(user){
	return {
		type: types.SWITCH_USER_AUTH_STATUS,
		user
	};
}
export function submitDetails(details, callback, route) {
	//route represents the endpoint to send this details
	return (dispatch) => {
		dispatch(startSubmitting());
		return axios.post(`${URL}${route}`, {
			'email': details['email'],
			'name': details['username'],
			'password': details['password']
		}
		).then((data) => {
			if (data.data.token !== undefined){
				//Response has token its a login i.e save the token
				localStorage.setItem('jwt',data.data.token);
				dispatch(setAuthStatusOfUser(jwt.decode(data.data.token)));
			}
			callback();
			dispatch(submissionSuccessful(data.data.message));
		}).catch((error) => {
			if(error.response != null) {
				return dispatch(submissionFailed(error.response.data.message));
			}
			if(!error.status){
				//network error
				return dispatch(submissionFailed(error.message));
			}
		});

	};
}
function userLoggedOut(message){
	return{
		type: types.SHOW_SNACK_BAR,
		message
	};
}
export function Logout(){
	return(dispatch)=>{
		localStorage.removeItem('jwt');
		dispatch(setAuthStatusOfUser({}));
		dispatch(userLoggedOut('You have Been Logged out.'));
	};
}