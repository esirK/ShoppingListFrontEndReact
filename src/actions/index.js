import axios from 'axios';

import * as types from '../constants/actiontypes';

export * from './shoppinglist_actions';
export * from './authentication_actions';
export * from './shoppinglistitems_actions';

export const URL = 'http://127.0.0.1:5000/v1/';


export function resetErrors() {
	return{
		type: types.CLEAR_ERRORS
	};
}
export function startSubmitting(){
	console.log('Started Submitting....');
	return {
		type: types.SUBMITTING_STARTED
	};
}
export function errorEncountered(error){
	return{
		type: types.ERROR_ENCOUNTERED,
		error
	};
}

export function hideSnackBar(dispatch){
	return(dispatch)=>{
		dispatch({type: types.HIDE_SNACK_BAR});
	};
}
/**
 * ShoppingList items
 */

//View Selected Shoppinglist
export function viewShoppingList(id){
	console.log('Looking for the shoppinglist ', id);
	return (dispatch) =>{
		//clear all errors first
		dispatch(resetErrors());
		dispatch(loadingShoppingListItems());
		return axios({
			method: 'get',
			url: `${URL}${'shoppinglists/'}${id}`,
			auth: {
				username: localStorage.getItem('jwt'),
				password: ''
			}
		}).then((response)=>{
			console.log('Got catcha ', response, response.data);
			dispatch(shoppinglistItemsRecieved(response.data, id));
		}).catch(error=>{
			dispatch(loadingShoppingListItemsFailed(error.response.data.message));
			dispatch(errorEncountered(error.response.data.message));
		});
	};
}
export function loadingShoppingListItems(){
	console.log('Loading items dispatched');
	return {
		type: types.SHOPPINGLIST_ITEMS_LOADING,
	};
}
function loadingShoppingListItemsFailed(error){
	return {
		type: types.SHOPPINGLIST_ITEMS_FAILED,
		error
	};
}
export function shoppinglistItemsRecieved(data, id){
	return {
		type: types.SHOPPINGLIST_ITEMS_LOADED_SUCCESSFULLY,
		data, id
	};
}