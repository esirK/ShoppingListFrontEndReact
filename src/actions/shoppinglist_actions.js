import axios from 'axios';


import * as types from '../constants/actiontypes';
import {errorEncountered, URL, resetErrors, startSubmitting} from './index';
/**
 * ShoppingLists
 */

export function getShoppingLists(){
	return (dispatch) => {
		dispatch(getShoppingListStarted());
		return axios({
			method: 'get',
			url: `${URL}${'shoppinglists'}`,
			auth: {
				username: localStorage.getItem('jwt'),
				password: ''
			}
		})
			.then(function (response) {
				console.log('Shoppinglists recieved as',response);
				dispatch(shoppinglistsRecieved(response.data));
			})
			.catch(function (error) {
				console.log('wahhh', error);
				dispatch(errorEncountered(error.message));
			});

	};
}
//Add a new shopping list function
export function addNewShoppingList(details){
	return (dispatch) =>{
		//clear all errors first
		dispatch(resetErrors());
		//Notify that adding a shopping list action has been dispatched
		dispatch(addingShoppinglistStarted());
		return axios({
			method: 'post',
			url: `${URL}${'shoppinglists'}`,
			data: {
				name: details['name'],
				description: details['description']
			  },
			auth: {
				username: localStorage.getItem('jwt'),
				password: ''
			}
		})
			.then(function (response){
				//Dispatch shoppinglist created successfully
				console.log('Got response when adding', response.data);
				dispatch(shoppinglistCreated(response.data.data, response.data.message));
			})
			.catch(function (error){
				//Dispatch shoppinglist creation failed
				console.log(error);
				if(error.message){
					dispatch(errorEncountered(error.message));
				}
				if(error.response.data.message !== 'undefined'){
					console.log('Got errorx', error );
				    dispatch(errorEncountered(error.response.data.message));
				}
			});
	};
}
//Delete an existing shoppinglist function
export function deleteShoppingList(id){
	return (dispatch)=>{
		//clear all errors first
		dispatch(resetErrors());
		dispatch(startSubmitting);
		return axios({
			method: 'delete',
			url: `${URL}${'shoppinglists/'}${id}`,
			auth: {
				username: localStorage.getItem('jwt'),
				password: ''
			}
		}).then((response)=>{
			//Dispatch shoppinglist DELETION Successfully
			dispatch(shoppinglistDeleted(response.data.message, response.data.data));
			console.log('Got response', response.data.message);
		}).catch((error)=>{
			//Dispatch shoppinglist DELETION Failed
			dispatch(errorEncountered(error.message));
		});
	};
}

export function updateShoppingList(id, details){
	//Updates an existing shoppinglist
	return (dispatch) =>{
		//clear all errors first
		dispatch(resetErrors());
		dispatch(updatingShoppinglistStarted());
		return axios({
			method: 'put',
			url: `${URL}${'shoppinglists/'}${id}`,
			data: {
				new_name: details['new_name'],
				description: details['description']
			  },
			auth: {
				username: localStorage.getItem('jwt'),
				password: ''
			}
		}).then((response)=>{
			if((response.data.message).includes('Nothing was provided')){
				dispatch(errorEncountered(response.data.message));
			}else{
				dispatch(closeUpdateDialog());
				dispatch(shoppinglistUpdated(response.data.message, response.data.data));
			}
		}).catch((error)=>{
			console.log('Got Meso', error);
			dispatch(errorEncountered(error.message));
		});
	};
}
function getShoppingListStarted(){
	console.log('Getting started');
	return{
		type: types.GETTING_SHOPPINGLISTS_STARTED
	};
}
function shoppinglistsRecieved(response){
	return{
		type: types.SHOPPINGLISTS_LOADED_SUCCESSFULLY,
		response
	};
}

function addingShoppinglistStarted(){
	return {
		type: types.CREATING_SHOPPINGLISTS_STARTED,
	};
}

function shoppinglistCreated(response, message){
	return {
		type: types.SHOPPINGLISTS_CREATED_SUCCESSFULY,
		response, message
	};
}
function shoppinglistDeleted(message, response){
	return {
		type: types.SHOPPINGLISTS_DELETED_SUCCESSFULY,
		message, response
	};
}
function updatingShoppinglistStarted(){
	return {
		type: types.UPDATING_SHOPPINGLIST_STARTED,
	};
}
function shoppinglistUpdated(message, response){
	return {
		type: types.SHOPPINGLISTS_UPDATED_SUCCESSFULY,
		message, response
	};
}
export function activateFab(){
	return(dispatch)=>{
		dispatch(
		 {type: types.ACTIVATE_FAB}
		);
	};
}
export function closeFab(){
	return(dispatch)=>{
		dispatch(
		 {type: types.DEACTIVATE_FAB}
		);
	};
}
export function openUpdateDialog(){
	return(dispatch)=>{
		dispatch({type: types.OPEN_UPDATE_DIALOG});
	};
}
export function closeUpdateDialog(){
	resetErrors();
	return(dispatch)=>{
		dispatch(
			{type: types.CLOSE_UPDATE_DIALOG}
		   );
	};
}