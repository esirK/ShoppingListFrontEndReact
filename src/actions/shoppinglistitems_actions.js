import axios from 'axios';

import * as types from '../constants/actiontypes';

import {errorEncountered, URL, resetErrors} from './index';

export function activateAddItem(){
	//Sets state of add item dialog to true
	return(dispatch)=>{
		dispatch(
			{type: types.ACTIVATE_ADD_ITEM}
		);
	};
}
export function deactivateAddItem(){
	//Sets state of add item dialog to false
	return(dispatch)=>{
		dispatch(resetErrors());
		dispatch(
			{type: types.DEACTIVATE_ADD_ITEM}
		);
	};
}
export function addNewShoppingListItem(details){
	return (dispatch) => {
		console.log('see whst you doing ', details);
		//clear all errors first
		dispatch(resetErrors());
		dispatch(addingShoppinglistItemStarted());
		return axios({
			method: 'post',
			url: `${URL}${'shoppinglist_items'}`,
			data:{
				name: details['name'].trim(),
				price: details['price'].trim(),
				quantity: details['quantity'].trim(),
				shoppinglist_id: details['shoppinglist_id']
			},
			auth: {
				username: localStorage.getItem('jwt'),
				password: ''
			}
		})
			.then(function (response) {
				console.log('Here is response of adding a shoppinglist item ', response.data);
				dispatch(shoppinglistItemCreated(response.data.message));
			})
			.catch(function (error) {
				console.log('wahhh', error.response);
				if(error.response.data.error === undefined){
					dispatch(errorEncountered(error.response.data.message));
				}else{
					dispatch(errorEncountered(error.response.data.error));
				}
			});

	};
}
export function deleteShoppinglistItem(id){
	return(dispatch)=>{
		dispatch(deletingItemStarted());
		return axios(
			{
				method: 'delete',
				url: `${URL}${'shoppinglist_items/'}${id}`,
				auth: {
					username: localStorage.getItem('jwt'),
					password: ''
				}
			}
		).then(function(response)
		{
			console.log('Got this response on deleting an item', response);
			dispatch(itemDeletedSuccessfully(response.data.message));
		}
		).catch(function(error){
			console.log('Error while trying to delete ', error);
		});
	};
}
function addingShoppinglistItemStarted(){
	return{
		type: types.ADDING_SHOPPINGLIST_ITEM_STARTED
	};
}
function shoppinglistItemCreated(message){
	return{
		type: types.SHOPPINGLIST_ITEM_ADDED, message
	};
}
function deletingItemStarted(){
	return{
		type: types.DELETING_ITEM_STARTED,
	};
}
function itemDeletedSuccessfully(message){
	return{
		type: types.ITEM_DELETED_SUCCESSFULLY,
		message
	};
}