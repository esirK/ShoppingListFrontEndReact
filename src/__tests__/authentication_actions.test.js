import React from 'react';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import moxios from 'moxios';
import axios from 'axios';
import * as actions from '../actions';
import * as types from '../constants/actiontypes';
import fetchMock from 'fetch-mock';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

global.window = {};
import localStorage from 'mock-local-storage';
window.localStorage = global.localStorage;

const store = mockStore();

beforeEach( () => {
	moxios.install();
});

afterEach( () => {
	moxios.uninstall();
});

describe('authentication actions', ()=>{
	it('should create an action to start loggin ', ()=>{
		const expectedAction = {
			type: types.SUBMITTING_STARTED
		};
		expect(actions.startSubmitting()).toEqual(expectedAction);
	});

	it('should return success message on successful submission', ()=>{
		const message = 'Details Submitted successfully';
		const expectedAction = {
			type: types.SUBMISSION_SUCCESSFUL, 
			message
		};
		expect(actions.submissionSuccessful('Details Submitted successfully')).toEqual(expectedAction);
	});
	it('should return error message on failure to submit', ()=>{
		const error = 'Network error while submitting';
		const expectedAction = {
			type: types.SUBMISSION_ERROR, 
			error
		};
		expect(actions.submissionFailed('Network error while submitting')).toEqual(expectedAction);
	});
});