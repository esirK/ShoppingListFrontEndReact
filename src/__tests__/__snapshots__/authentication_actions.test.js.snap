// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`submitDetails actions Returns user details when login is successful 1`] = `
Array [
  Object {
    "type": "SUBMITTING_STARTED",
  },
  Object {
    "type": "SWITCH_USER_AUTH_STATUS",
    "user": Object {
      "expires_at": 1514794203165,
      "id": 27,
      "username": "esir",
    },
  },
  Object {
    "message": undefined,
    "type": "SUBMISSION_SUCCESSFUL",
  },
]
`;
